src/wallet/coin_selection.rs

```diff
@@ -138,6 +145,102 @@ impl CoinSelectionResult {
     }
 }
 
+/// Metric introduced to measure the performance of different coin selection algorithms.
+///
+/// This implementation considers "waste" the sum of two values:
+/// * Timing cost
+/// * Creation cost
+/// > waste = timing_cost + creation_cost
+///
+/// **Timing cost** is the cost associated with the current fee rate and some long term fee rate used
+/// as a threshold to consolidate UTXOs.
+/// > timing_cost = txin_size * current_fee_rate - txin_size * long_term_fee_rate
+///
+/// Timing cost can be negative if the `current_fee_rate` is cheaper than the `long_term_fee_rate`,
+/// or zero if they are equal.
+///
+/// **Creation cost** is the cost associated with the surplus of coins beyond the transaction amount
+/// and transaction fees. It can appear in the form of a change output or in the form of excess
+/// fees paid to the miner.
+///
+/// Change cost is derived from the cost of adding the extra output to the transaction and spending
+/// that output in the future.
+/// > change_cost = current_fee_rate * change_output_size + long_term_feerate * change_spend_size
+///
+/// Excess happens when there is no change, and the surplus of coins is spend as part of the fees
+/// to the miner:
+/// > excess = tx_total_value - tx_fees - target
+///
+/// Where _target_ is the desired amount to send.
+///
+/// Creation cost can be zero if there is a perfect match as result of the coin selection
+/// algorithm.
+///
+/// So, waste can be zero if creation and timing cost are zero. Or can be negative, if timing cost
+/// is negative and the creation cost is low enough (less than the absolute value of timing
+/// cost).
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Waste(pub i64);
+
+const LONG_TERM_FEE_RATE: FeeRate = FeeRate::from_sat_per_vb(5.0);
+
+impl Waste {
+    /// Calculate the amount of waste for the given coin selection
+    ///
+    /// - `selected`: the selected transaction inputs
+    /// - `cost_of_change`: the cost of creating change and spending it in the future
+    /// - `target`: threshold in satoshis used to select UTXOs. It includes the sum of recipient
+    ///             outputs, the fees for creating the recipient outputs and the fees for fixed transaction
+    ///             parts
+
+    /// - `fee_rate`: fee rate to use
+    pub fn calculate(
+        selected: &[WeightedUtxo],
+        cost_of_change: Option<u64>,
```

Comment:
Really picky nit: I'd stay consistent between `change_cost` and `cost_of_change`, as far as I understand they're the same thing, so we should either use one name or the other in the code (whichever you find more clear)

Reply:
Btw, probably the cost of change should be calculated inside the `Waste::calculate` and not outside. 

> change_cost = current_fee_rate * change_output_size + long_term_feerate * change_spend_size

This means we need the following:
`current_fee_rate` -> we have it already
`change_output_size` -> we need it to have it as a param
`long_term_feerate` -> we have it, it's `LONG_TERM_FEE_RATE`
`change_spend_size` -> we need it to have it as a param

Which means we should either add two new parameters to the function (maybe it's better to merge those in a struct)

Reply:
As of exactly how to calculate the `change_output_size` and the `change_spend_size`, I'm figuring it out, and I'll get back to you